<a href="https://colab.research.google.com/github/mehrnazh/PythonVisualization/blob/main/My_first_notebook.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# First Steps with Python and Jupyter





In [ ]:
# @title Tutorial slides
# @markdown These are the slides for this tutorial
from IPython.display import HTML
sway_id = "https://sway.cloud.microsoft/Jylr4v2KlnSxVbRm?ref=Link"

sway_embed_code = f"""
<iframe src="https://sway.cloud.microsoft/s/Jylr4v2KlnSxVbRm/embed" width="760" height="500" frameborder="0" marginwidth="0" marginheight="0" style="border: none; max-width: 100%; max-height: 100vh;" allowfullscreen mozallowfullscreen msallowfullscreen webkitallowfullscreen></iframe>
"""

print(sway_embed_code)
HTML(sway_embed_code)


<iframe src="https://sway.cloud.microsoft/s/Jylr4v2KlnSxVbRm/embed" width="760" height="500" frameborder="0" marginwidth="0" marginheight="0" style="border: none; max-width: 100%; max-height: 100vh;" allowfullscreen mozallowfullscreen msallowfullscreen webkitallowfullscreen></iframe>



This tutorial covers the following topics:

* Start running codes in Python using google colab
* Performing arithmetic operations using Python
* Solving multi-step problems using variables
* Evaluating conditions using Python
* Combining conditions with logical operators
* Adding text styles using Markdown

### How to run the code

This tutorial is an executable [Jupyter notebook](https://jupyter.org) hosted on [Google Colab](https://colab.research.google.com/) (don't worry if these terms seem unfamiliar; we'll learn more about them soon). You can _run_ this tutorial and experiment with the code examples in a couple of ways: *using free online resources* (recommended) or *on your computer*.

#### Option 1: Running using free online resources (1-click, recommended)

The easiest way to start executing the code is to click the **Run** button of code cells. you'll need to create an account on [Google Colab](https://colab.research.google.com) to use this platform. It's free.


#### Option 2: Running on your computer locally

To run the code on your computer locally, you'll need to set up [Python](https://www.python.org), download the notebook and install the required libraries.

>  **Jupyter Notebooks**: This tutorial is a [Jupyter notebook](https://jupyter.org) - a document made of _cells_. Each cell can contain code written in Python or explanations in plain English. You can execute code cells and view the results, e.g., numbers, messages, graphs, tables, files, etc. instantly within the notebook. Jupyter is a powerful platform for experimentation and analysis. Don't be afraid to mess around with the code & break things - you'll learn a lot by encountering and fixing errors. You can use the "Kernel > Restart & Clear Output" menu option to clear all outputs and start again from the top.

## Run the tradition of printing Hello World! in the cell below

## Performing Arithmetic Operations using Python

Let's begin by using Python as a calculator. You can write and execute Python using a code cell within Jupyter.

> **Working with cells**: To create a new cell within Colab, you can simply insert code cell... You can also use the keyboard shortcut `Esc+B` to create a new cell.

Run the code cells below to perform calculations and view their result. Try changing the numbers and run the modified cells again to see updated results. Can you guess what the `//`, `%`, and `**` operators are used for?

In [ ]:
2 + 3 + 9

14

In [ ]:
99 - 73

26

In [ ]:
23.54 * -1432

-33709.28

In [ ]:
100 / 7

14.285714285714286

In [ ]:
100 // 7  #Oh wait! what is this operation?

14

In [ ]:
100 % 7  #Oh wait! what about this?

2

In [ ]:
5 ** 3

125

As you might expect, operators like `/` and `*` take precedence over other operators like `+` and `-` as per mathematical conventions. You can use parentheses, i.e. `(` and `)`, to specify the order in which operations are performed.

In [ ]:
((2 + 5) * (17 - 3)) / (4 ** 3)

1.53125

Python supports the following arithmetic operators:

| Operator   | Purpose           | Example     | Result    |
|------------|-------------------|-------------|-----------|
| `+`        | Addition          | `2 + 3`     | `5`       |
| `-`        | Subtraction       | `3 - 2`     | `1`       |
| `*`        | Multiplication    | `8 * 12`    | `96`      |
| `/`        | Division          | `100 / 7`   | `14.28..` |
| `//`       | Floor Division    | `100 // 7`  | `14`      |    
| `%`        | Modulus/Remainder | `100 % 7`   | `2`       |
| `**`       | Exponent          | `5 ** 3`    | `125`     |


Try solving some simple problems, Solve the following:

1. The population of a town is 198568. Out of them 45312 are men and 35678 are women. Find the number of children in the town.

2. A shopkeeper has 2425 boxes of 24 pencils each. How many pencils do all the boxes have in all?

3. Linda bought a coat for 2265 dollars and a saree for 2150 dollars. She gave $5000 to the shopkeeper. How much money did the shopkeeper return to her?

4. The cost of 21 TV sets is $95844. Find the cost of one TV set.

5. A factory produces 24532 bulbs in a month. What is its annual production?
You can use the empty cells below and add more cells if required.

## Solving multi-step problems using variables

Let's try solving the following word problem using Python:

> A grocery store sells a bag of ice for $1.25 and makes a 20% profit. If it sells 500 bags of ice, how much total profit does it make?

We can list out the information provided and gradually convert the word problem into a mathematical expression that can be evaluated using Python.

*Cost of ice bag ($)* = 1.25

*Profit margin* = 20% = .2

*Profit per bag ($)* = profit margin * cost of ice bag = .2 * 1.25

*No. of bags* = 500

*Total profit* = no. of bags * profit per bag = 500 * (.2 * 1.25)

In [ ]:
500 * (.2 * 1.25)

125.0

Thus, the grocery store makes a total profit of $125. While this is a reasonable way to solve a problem, it's not entirely clear by looking at the code cell what the numbers represent. We can give names to each of the numbers by creating Python *variables*.

> **Variables**: While working with a programming language such as Python, information is stored in *variables*. You can think of variables as containers for storing data. The data stored within a variable is called its *value*.

In [ ]:
cost_of_ice_bag = 1.25

In [ ]:
profit_margin = .2

In [ ]:
number_of_bags = 500

The variables `cost_of_ice_bag`, `profit_margin`, and `number_of_bags` now contain the information provided in the word problem. We can check the value of a variable by typing its name into a cell. We can combine variables using arithmetic operations to create other variables.

> **Code completion**: While typing the name of an existing variable in a code cell within Jupyter, just type the first few characters and press the `Tab` key to autocomplete the variable's name. Try typing `pro` in a code cell below and press `Tab` to autocomplete to `profit_margin`.

In [ ]:
# get profit margin in this code

In [ ]:
# calculate the profit_per_bag and store the value as a new variable
profit_per_bag =

In [ ]:
profit_per_bag

0.25

In [ ]:
total_profit =

In [ ]:
total_profit

125.0

If you try to view the value of a variable that has not been *defined*, i.e., given a value using the assignment statement `variable_name = value`, Python shows an error.

In [ ]:
net_profit

NameError: name 'net_profit' is not defined

Think!
what are variable type of the following:


`int = 3` , `float = 3.0` , `str = 'My Name is Mehrnaz'`

use the code cell below to test




Storing and manipulating data using appropriately named variables is a great way to explain what your code does.

Let's display the result of the word problem using a friendly message. We can do this using the `print` *function*.

> **Functions**: A function is a reusable set of instructions. It takes one or more inputs, performs certain operations, and often returns an output. Python provides many in-built functions like `print` and also allows us to define our own functions.



In [ ]:
print("The grocery store makes a total profit of $", total_profit)

The grocery store makes a total profit of $ 125.0


> **`print`**: The `print` function is used to display information. It takes one or more inputs, which can be text (within quotes, e.g., `"this is some text"`), numbers, variables, mathematical expressions, etc. We'll learn more about variables & functions in the next tutorial.

> list of Python's built-in functions can be found here: [Python's built-in functions](https://docs.python.org/3/library/functions.html)

Creating a code cell for each variable or mathematical operation can get tedious. Fortunately, Jupyter allows you to write multiple lines of code within a single code cell. The result of the last line of code within the cell is displayed as the output.

Let's rewrite the solution to our word problem within a single cell.

In [ ]:
# Store input data in variables
cost_of_ice_bag = 1.25
profit_margin = .2
number_of_bags = 500

# Perform the required calculations
profit_per_bag = cost_of_ice_bag * profit_margin
total_profit = number_of_bags * profit_per_bag

# Display the result
print("The grocery store makes a total profit of $", total_profit)

The grocery store makes a total profit of $ 125.0


Note that we're using the `#` character to add *comments* within our code.

> **Comments**: Comments and blank lines are ignored during execution, but they are useful for providing information to humans (including yourself) about what the code does. Comments can be inline (at the end of some code), on a separate line, or even span multiple lines.

Inline and single-line comments start with `#`, whereas multi-line comments begin and end with three quotes, i.e. `"""`. Here are some examples of code comments:

In [ ]:
my_favorite_number = 1 # an inline comment

In [ ]:
# This comment gets its own line
my_least_favorite_number = 3

In [ ]:
"""This is a multi-line comment.
Write as little or as much as you'd like.

Comments are really helpful for people reading
your code, but try to keep them short & to-the-point.

Also, if you use good variable names, then your code is
often self explanatory, and you may not even need comments!
"""
a_neutral_number = 5



> **EXERCISE**: A travel company wants to fly a plane to the Bahamas. Flying the plane costs 5000 dollars. So far, 29 people have signed up for the trip. If the company charges 200 dollars per ticket, what is the profit made by the company? Create variables for each numeric quantity and use appropriate arithmetic operations.

## More on ```print()``` function



### Print formatting
Print formatting is handy for displaying parameters in a clean and organized form. Python 3.6 introduced the new string formatting f-strings. Since we are dealing with type float variables, we use `f'{x:.3f}'` for formatting x to three decimal points, and `f'{x:.4e}'` for four decimal points but in exponential notation.

In [ ]:
x = 3.14159265e-1
print("the number is:", x)
print('\nLets see how we can format number printing:') # what does "\n" do?
print("the number with 3 decimal points is:", f'{x:.3f}') # note the comma here
print("the number with 4 decimal points in exponential notation is:\n "f'{x:.4e}')

the number is: 0.314159265

Lets see how we can format number printing:
the number with 3 decimal points is: 0.314
the number with 4 decimal points in exponential notation is:
 3.1416e-01


## Evaluating conditions using Python

Apart from arithmetic operations, Python also provides several operations for comparing numbers & variables.

| Operator    | Description                                                     |
|-------------|-----------------------------------------------------------------|
| `==`        | Check if operands are equal                                     |
| `!=`        | Check if operands are not equal                                 |
| `>`         | Check if left operand is greater than right operand             |
| `<`         | Check if left operand is less than right operand                |
| `>=`        | Check if left operand is greater than or equal to right operand |
| `<=`        | Check if left operand is less than or equal to right operand    |

The result of a comparison operation is either `True` or `False` (note the uppercase `T` and `F`). These are special keywords in Python. Let's try out some experiments with comparison operators.

In [ ]:
my_favorite_number = 1
my_least_favorite_number = 5
a_neutral_number = 3

In [ ]:
# Equality check - True
my_favorite_number == 1

In [ ]:
# Equality check - False
my_favorite_number == my_least_favorite_number

In [ ]:
# Not equal check - True
my_favorite_number != a_neutral_number

In [ ]:
# Not equal check - False
a_neutral_number != 3

In [ ]:
# Greater than check - True
my_least_favorite_number > a_neutral_number

In [ ]:
# Greater than check - False
my_favorite_number > my_least_favorite_number

In [ ]:
# Less than check - True
my_favorite_number < 10

In [ ]:
# Less than check - False
my_least_favorite_number < my_favorite_number

In [ ]:
# Greater than or equal check - True
my_favorite_number >= 1

In [ ]:
# Greater than or equal check - False
my_favorite_number >= 3

In [ ]:
# Less than or equal check - True
3 + 6 <= 9

In [ ]:
# Less than or equal check - False
my_favorite_number + a_neutral_number <= 3

Just like arithmetic operations, the result of a comparison operation can also be stored in a variable.

In [ ]:
cost_of_ice_bag = 1.25
is_ice_bag_expensive = cost_of_ice_bag >= 10
print("Is the ice bag expensive?", is_ice_bag_expensive)

## Combining conditions with logical operators

The logical operators `and`, `or` and `not` operate upon conditions and `True` & `False` values (also known as *booleans*). `and` and `or` operate on two conditions, whereas `not` operates on a single condition.

The `and` operator returns `True` when both the conditions evaluate to `True`. Otherwise, it returns `False`.

| `a`     | `b`    | `a and b` |
|---------|--------|-----------|
|  `True` | `True` | `True`    |
|  `True` | `False`| `False`   |
|  `False`| `True` | `False`   |
|  `False`| `False`| `False`   |


In [ ]:
my_favorite_number

In [ ]:
my_favorite_number > 0 and my_favorite_number <= 3

In [ ]:
my_favorite_number < 0 and my_favorite_number <= 3

In [ ]:
my_favorite_number > 0 and my_favorite_number >= 3

In [ ]:
True and False

In [ ]:
True and True

The `or` operator returns `True` if at least one of the conditions evaluates to `True`. It returns `False` only if both conditions are `False`.

| `a`     | `b`    | `a or b`  |
|---------|--------|-----------|
|  `True` | `True` | `True`    |
|  `True` | `False`| `True`    |
|  `False`| `True` | `True`    |
|  `False`| `False`| `False`   |



In [ ]:
a_neutral_number = 3

In [ ]:
a_neutral_number == 3 or my_favorite_number < 0

In [ ]:
a_neutral_number != 3 or my_favorite_number < 0

In [ ]:
my_favorite_number < 0 or True

In [ ]:
False or False

The `not` operator returns `False` if a condition is `True`, and `True` if the condition is `False`.

In [ ]:
not a_neutral_number == 3

In [ ]:
not my_favorite_number < 0

In [ ]:
not False

In [ ]:
not True

Logical operators can be combined to form complex conditions. Use round brackets or parentheses `(` and `)` to indicate the order in which logical operators should be applied.

In [ ]:
(2 > 3 and 4 <= 5) or not (my_favorite_number < 0 and True)

In [ ]:
not (True and 0 < 1) or (False and True)

If parentheses are not used, logical operators are applied from left to right.

In [ ]:
not True and 0 < 1 or False and True

Experiment with arithmetic, conditional and logical operators in Python using the interactive nature of Jupyter notebooks. We will learn more about variables and functions in future tutorials.

## Adding text styles using Markdown

Adding explanations using text cells (like this one) is a great way to make your notebook informative for other readers. It is also useful if you need to refer back to it in the future. You can double click on a text cell within Jupyter to edit it. In the edit mode, you'll notice that the text looks slightly different (for instance, the heading has a `##` prefix. This text is written using Markdown, a simple way to add styles to your text. Execute this cell to see the output without the special characters. You can switch back and forth between the source and the output to apply a specific style.

For instance, you can use one or more `#` characters at the start of a line to create headers of different sizes:

# Header 1

## Header 2

### Header 3

#### Header 4

To create a bulleted or numbered list, simply start a line with `*` or `1.`.

A bulleted list:

* Item 1
* Item 2
* Item 3

A numbered list:

1. Apple
2. Banana
3. Pineapple

You can make some text bold using `**`, e.g., **some bold text**, or make it italic using `*`, e.g., *some italic text.* You can also create links, e.g., [a link](https://scholar.google.com/). Images are easily embedded too:


![embed](<iframe width="760px" height="500px" src="https://sway.cloud.microsoft/s/Jylr4v2KlnSxVbRm/embed" frameborder="0" marginheight="0" marginwidth="0" max-width="100%" sandbox="allow-forms allow-modals allow-orientation-lock allow-popups allow-same-origin allow-scripts" scrolling="no" style="border: none; max-width: 100%; max-height: 100vh" allowfullscreen mozallowfullscreen msallowfullscreen webkitallowfullscreen></iframe>)

Another really nice feature of Markdown is the ability to include blocks of code. Note that code blocks inside Markdown cells cannot be executed.

```
# Perform the required calculations
profit_per_bag = cost_of_ice_bag * profit_margin
total_profit = number_of_bags * profit_per_bag

# Display the result
print("The grocery store makes a total profit of $", total_profit)

```

You can learn the full syntax of Markdown here: https://learnxinyminutes.com/docs/markdown/

## Save and upload your notebook

Whether you're running this Jupyter notebook online or on your computer, it's essential to save your work from time to time. You can continue working on a saved notebook later or share it with friends and colleagues to let them execute your code. [GitHub](https://github.com/) offers an easy way of saving and sharing your Jupyter notebooks online.

**Step 1:**
Create a GitHub Repository:

Go to GitHub, sign in, and create a new repository. Copy the repository URL.

**Step 2:** Set Up Git Configuration:






In [16]:
!git config --global user.email "your_email@example.com"
!git config --global user.name "Your username"

**Step 3:** Generate a Personal Access Token:

* Go to GitHub Settings > Developer settings > Personal access tokens > Tokens (classic).
* Click "Generate new token," select scopes (e.g., repo), and generate the token.
Copy the token.
Open Google Colab:

**Step 3:** Clone GitHub Repository in Colab

In [ ]:
from getpass import getpass
import os

# Prompt for GitHub username and token
username = input('GitHub Username: ')
token = getpass('GitHub Token: ')

# Construct GitHub URL with token for authentication
repo_url = f'https://{username}:{token}@github.com/YOUR_GITHUB_USERNAME/YOUR_REPO_NAME.git'

# Clone the repository
!git clone {repo_url}


**Step 4:** Navigate to the Repository Directory:

In [ ]:
%cd YOUR_REPO_NAME

**Step 5:** Copy the Notebook to the Repository Directory:

* Copy the colab Notebook into your Google drive
* Mount Google drive into Colab
* Copy file from your Google Drive to your Repository Directory

In [8]:
from google.colab import drive
drive.mount('/content/drive')  # you need to give necessary access

Mounted at /content/drive


In [10]:
#Copy the Notebook to the Repository Directory:
# Replace 'your_notebook.ipynb' with the name of your notebook

!cp /content/your_notebook.ipynb /content/YOUR_REPO_NAME/

**Step 6:** Stage, Commit, and Push the Notebook to GitHub:

In [19]:
# Check repository status
!git status

On branch main
Your branch is up to date with 'origin/main'.

nothing to commit, working tree clean


In [22]:
# Stage all changes
!git add .

In [ ]:
# Stage the new notebook file
!git add your_notebook.ipynb

Finally, you can run `commit` to
capture and upload a snapshot of the notebook.

In [23]:
# Commit changes with a message
!git commit -m "Add Colab NoteBook"

On branch main
Your branch is up to date with 'origin/main'.

nothing to commit, working tree clean


In [24]:
# Push changes to the main branch
!git push origin main

Everything up-to-date


## Further Reading and References

Following are some resources where you can learn about more arithmetic, conditional and logical operations in Python:

* Python Tutorial at W3Schools: https://www.w3schools.com/python/
* Practical Python Programming: https://dabeaz-course.github.io/practical-python/Notes/Contents.html
* Python official documentation: https://docs.python.org/3/tutorial/index.html

Now that you have taken your first steps with Python, you are ready to move on to the next tutorial: ["A Quick Tour of Variables and Data Types in Python"](https://jovian.ml/aakashns/python-variables-and-data-types).

## Questions for Revision

Try answering the following questions to test your understanding of the topics covered in this notebook:

1. What is a Jupyter notebook?
2. How do you add a new code cell below an existing cell?
3. How do you add a new Markdown cell below an existing cell?
4. How do you convert a code cell to a Markdown cell or vice versa?
5. How do you execute a code cell within Jupyter?
6. What the different arithmetic operations supported in Python?
7. How do you perform arithmetic operations using Python?
8. What is the difference between the `/` and the `//` operators?
9. What is the difference between the `*` and the `**` operators?
10. What is the order of precedence for arithmetic operators in Python?
11. How do you specify the order in which arithmetic operations are performed in an expression involving multiple operators?
12. How do you solve a multi-step arithmetic word problem using Python?
13. What are variables? Why are they useful?
14. How do you create a variable in Python?
15. What is the assignment operator in Python?
16. What are the rules for naming a variable in Python?
17. How do you view the value of a variable?
18. How do you store the result of an arithmetic expression in a variable?
19. What happens if you try to access a variable that has not been defined?
20. How do you display messages in Python?
21. What type of inputs can the print function accept?
22. What are code comments? How are they useful?
23. What are the different ways of creating comments in Python code?
24. What are the different comparison operations supported in Python?
25. What is the result of a comparison operation?
26. What is the difference between `=` and `==` in Python?
27. What are the logical operators supported in Python?
28. What is the difference between the `and` and `or` operators?
29. Can you use comparison and logical operators in the same expression?
30. What is the purpose of using parentheses in arithmetic or logical expressions?
31. What is Markdown? Why is it useful?
32. How do you create headings of different sizes using Markdown?
33. How do you create bulleted and numbered lists using Markdown?
34. How do you create bold or italic text using Markdown?
35. How do you include links & images within Markdown cells?
36. How do you include code blocks within Markdown cells?
37. Is it possible to execute the code blocks within Markdown cells?
38. How do you upload and share your Jupyter notebook online using Jovian?
39. What is the purpose of the API key requested by jovian.commit ? Where can you find the API key?
40. Where can you learn about arithmetic, conditional and logical operations in Python?


## Solution for Exercise



> **EXERCISE**: A travel company wants to fly a plane to the Bahamas. Flying the plane costs 5000 dollars. So far, 29 people have signed up for the trip. If the company charges 200 dollars per ticket, what is the profit made by the company? Create variables for each numeric quantity and use appropriate arithmetic operations.

In [ ]:
plane_cost=5000
total_people_signed_up=29
ticket_cost=200

In [ ]:
amount_received_by_company=total_people_signed_up*ticket_cost

In [ ]:
total_profit=amount_received_by_company-plane_cost

In [ ]:
total_profit

800

The travel company is in a profit of $800!

## Git commands

These commands should cover most of your basic Git needs. For more advanced usage, you can refer to the [official Git documentation](https://git-scm.com/doc).

### Git Configuration

1. Set Up User Information:

In [ ]:
!git config --global user.name "Your Name"
!git config --global user.email "your_email@example.com"

2. Check Configuration Settings:

In [ ]:
!git config --list

### Repository Initialization
3. Initialize a New Repository:


In [ ]:
!git init

4. Clone an Existing Repository:

In [ ]:
!git clone https://github.com/user/repo.git

### Staging and Committing
5. Check Repository Status:



In [ ]:
!git status

6. Add Files to Staging Area:



In [ ]:
!git add file_name

# Add all files
!git add .

7. Commit Changes:


In [ ]:
!git commit -m "Commit message"

### Branching and Merging
8. List All Branches:

In [ ]:
!git branch

9. Create a New Branch:

In [ ]:
!git branch branch_name

10. Switch to a Branch:

In [ ]:
!git checkout branch_name

11. Create and Switch to a New Branch:

In [ ]:
!git checkout -b branch_name

12. Merge a Branch into Current Branch:

In [ ]:
!git merge branch_name

13. Delete a Branch:

In [ ]:
!git branch -d branch_name

### Remote Repositories
14. Add a Remote Repository:

In [ ]:
!git remote add origin https://github.com/user/repo.git

15. View Remote Repositories:


In [ ]:
!git remote -v

16. Push Changes to Remote Repository:

In [ ]:
!git push origin branch_name

17. Pull Changes from Remote Repository:

In [ ]:
!git pull origin branch_name

18. Fetch Changes from Remote Repository:

In [ ]:
!git fetch origin

### Viewing History
19. View Commit History:

In [ ]:
!git log

20. View a Brief Commit History:

In [ ]:
!git log --oneline

21. View Changes (Diff):

In [ ]:
!git diff

### Undoing Changes
22. Unstage a File:

In [ ]:
!git reset HEAD file_name

23. Revert Changes in a File:

In [ ]:
!git checkout -- file_name

24. Amend Last Commit:

In [ ]:
!git commit --amend

### Tagging
25. Create a New Tag:

In [ ]:
!git tag tag_name

26. Push a Tag to Remote Repository:


In [ ]:
!git push origin tag_name

27. List Tags:

In [ ]:
!git tag